<a href="https://colab.research.google.com/github/NadezhdaGruzdeva/Statistics-Probability/blob/main/statistics_probability_hw2_bernully_puasson_and_numpy_practice.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import math

In [ ]:
def Combination(k,n): 
  '''Функция для сочетний
  k - количество элементов в наборе
  n - общее число элементов
  '''
  return int(math.factorial(n) / (math.factorial(k) * math.factorial(n - k)))

In [ ]:
def Razmeshenie(k,n): 
  '''Функция для размещения
  k - количество элементов в наборе
  n - общее число элементов
  '''
  return int(math.factorial(n) / math.factorial(n - k))

In [ ]:
def Bernully(k,n,p):
  '''P(X=k) = C_n ^k    *〖p 〗^k*q^(n-k)
позволяющая находить вероятность появления события 
A определённое количество раз при любом числе независимых испытаний
	k - число наступления события (дискретная величина из отрезка [0, n])
	n -  число испытаний 
	p - вероятность наступления события A в  независимых испытаниях, 
	q = 1 – p. Противоположная вероятность

  '''
  return Combination(k,n) * p ** k * (1 - p) ** (n - k)

In [ ]:
def Puasson(k,n,p):
  '''
P(k)≈λ^k/k!*e^(-λ)

k — количество событий,
λ  — математическое ожидание случайной величины (среднее количество событий за фиксированный промежуток времени),
k! обозначает факториал числа 
e=2,718.. — основание натурального логарифма
вероятность, что событие наступит m раз из n

λ. Это средняя интенсивность наступления события на некоторую единицу измерения
λ = p*k 

Распределе́ние Пуассо́на — распределение дискретного типа случайной величины, представляющей собой число событий, 
произошедших за фиксированное время, при условии, что данные события происходят с некоторой фиксированной средней 
интенсивностью и независимо друг от друга.

  '''
  λ = p*n
  return (λ) ** k / math.factorial(k) * math.exp(-λ)

1) Вероятность того, что стрелок попадет в мишень, выстрелив один раз, равна 0.8. Стрелок выстрелил 100 раз. Найдите вероятность того, что стрелок попадет в цель ровно 85 раз.

In [ ]:
Bernully(85,100,0.8)

0.048061793700746355

Ответ: 4,8%

2) Вероятность того, что лампочка перегорит в течение первого дня эксплуатации, равна 0.0004. В жилом комплексе после ремонта в один день включили 5000 новых лампочек. Какова вероятность, что ни одна из них не перегорит в первый день? Какова вероятность, что перегорят ровно две?

In [ ]:
Puasson(0,5000,0.0004)

0.1353352832366127

In [ ]:
Puasson(2,5000,0.0004)

0.2706705664732254

Ответ: 13,5% и 27% соответственно

3) Монету подбросили 144 раза. Какова вероятность, что орел выпадет ровно 70 раз?

In [ ]:
Bernully(70,144,0.5)

0.06281178035144776

Ответ: 6,3% 

4) В первом ящике находится 10 мячей, из которых 7 - белые. Во втором ящике - 11 мячей, из которых 9 белых. Из каждого ящика вытаскивают случайным образом по два мяча. Какова вероятность того, что все мячи белые? Какова вероятность того, что ровно два мяча белые? Какова вероятность того, что хотя бы один мяч белый?

In [ ]:
Bernully(2,2,0.7)*Bernully(2,2,9/11) #for 4 whiteballs (all)

0.32801652892561983

In [ ]:
Bernully(2,2,0.7)*Bernully(0,2,9/11) + Bernully(0,2,0.7)*Bernully(2,2,9/11) + Bernully(1,2,0.7)*Bernully(1,2,9/11) #for 2 white balls

0.20140495867768599

In [ ]:
zeroWhiteBalls = Bernully(0,2,0.7)*Bernully(0,2,9/11) #for 0 white balls
zeroWhiteBalls

0.002975206611570247

In [ ]:
1 - zeroWhiteBalls #at least 1 white ball

0.9970247933884298

Ответы: 32,8%, 20.1%, 99.7%

Эмпирическое доказательство на примере 1 задачи

1) Вероятность того, что стрелок попадет в мишень, выстрелив один раз, равна 0.8. Стрелок выстрелил 100 раз. Найдите вероятность того, что стрелок попадет в цель ровно 85 раз.

In [ ]:
import numpy as np
a = np.ones(800)
a

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1.

In [ ]:
b =np.zeros(200)
b

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

In [ ]:
full = np.hstack([a, b])
full

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1.

In [ ]:
np.random.shuffle(full)
full

array([1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0.,
       0., 1., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 0., 1., 1., 1.,
       0., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 0., 1., 1., 1., 1.,
       1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1.,
       0., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 1., 0., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 0., 1., 1.,
       1., 0., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1.,
       0., 1., 1., 0., 1., 1., 0., 0., 0., 0., 1., 0., 1., 1., 0., 1., 1.,
       1., 0., 0., 1., 1., 1., 1., 0., 1., 1., 1., 1., 0., 1., 1., 1., 1.,
       0., 1., 0., 1., 1., 0., 1., 1., 1., 0., 1., 1., 1., 1., 1., 0., 0.,
       0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0.,
       1., 0., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1.,
       0., 1., 1., 1., 1.

In [ ]:
trialsData = full.copy()
trialsData


array([1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0.,
       0., 1., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 0., 1., 1., 1.,
       0., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 0., 1., 1., 1., 1.,
       1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1.,
       0., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 1., 0., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 0., 1., 1.,
       1., 0., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1.,
       0., 1., 1., 0., 1., 1., 0., 0., 0., 0., 1., 0., 1., 1., 0., 1., 1.,
       1., 0., 0., 1., 1., 1., 1., 0., 1., 1., 1., 1., 0., 1., 1., 1., 1.,
       0., 1., 0., 1., 1., 0., 1., 1., 1., 0., 1., 1., 1., 1., 1., 0., 0.,
       0., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0.,
       1., 0., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 1.,
       0., 1., 1., 1., 1.

In [ ]:
trialsQty = 9999

In [ ]:
for i in range (trialsQty):
  np.random.shuffle(full)
  trialsData = np.vstack([trialsData, full])

In [ ]:
np.shape(trialsData)

(10000, 1000)

In [ ]:

resultsTrials = np.sum(trialsData[:,:100], axis = 1)
np.shape(resultsTrials)

(10000,)

In [ ]:
np.shape(resultsTrials)

(10000,)

In [ ]:
resultsTrials85 = resultsTrials[resultsTrials == 85]
resultsTrials85

array([85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85.,
       85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85., 85

In [ ]:
np.shape(resultsTrials85)

(429,)

In [ ]:
import pandas as pd

In [ ]:
df = pd.DataFrame(resultsTrials)

In [ ]:
pivotResults=df.value_counts()
type(pivotResults)

pandas.core.series.Series

In [ ]:
pivotResults

80.0    1086
81.0    1005
79.0    1005
82.0     939
78.0     898
83.0     822
77.0     717
84.0     674
76.0     548
75.0     444
85.0     429
74.0     319
86.0     303
87.0     181
73.0     172
72.0     120
88.0     110
71.0      65
89.0      46
70.0      42
90.0      24
69.0      16
68.0      14
91.0       7
66.0       5
92.0       4
67.0       2
93.0       1
61.0       1
94.0       1
dtype: int64

In [ ]:
pivotResults.sort_index()

61.0       1
66.0       5
67.0       2
68.0      14
69.0      16
70.0      42
71.0      65
72.0     120
73.0     172
74.0     319
75.0     444
76.0     548
77.0     717
78.0     898
79.0    1005
80.0    1086
81.0    1005
82.0     939
83.0     822
84.0     674
85.0     429
86.0     303
87.0     181
88.0     110
89.0      46
90.0      24
91.0       7
92.0       4
93.0       1
94.0       1
dtype: int64

На 10тыс испытаний эмпиримческим путем выявлено, что вероятноть попадания 85 раз из 100 составляет 4,43%